### PR TITLE
Simplified usage of class AssociableToAST

### PR DESCRIPTION
**Pull request by 4everTheOne**
According to the documentation in [AssociableToAST#toAst()](https://www.javadoc.io/doc/com.github.javaparser/javaparser-core/latest/com/github/javaparser/resolution/declarations/AssociableToAST.html) all the Resolved declaration most be asociable to a AST. 

Currently this doesn't happen, making it hard do access a node from a ResolvedDeclaration. To work around this we need to make some check that requires a lot of boilerplate like in the example below.

https://github.com/javaparser/javaparser/blob/1ce7d85c2b1653a192169ea245c14c537a99132b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/CompilationUnitContext.java#L353-L366

This PR suggest the remove of the type parameter of the class AssociableToAST and use a helper method to convert to the desired type.
To make it more clear lets take a look at the example mentioned above but this time in the new format:
```java
private boolean isAncestorOf(ResolvedTypeDeclaration descendant) {
    return descendant.toAst()
            .filter(node -> wrappedNode.isAncestorOf(node))
            .isPresent();
}
``` 

Other example, previously we would have to do something like this:
```java
ResolvedDeclaration resolvedDeclaration = ...;
if (resolvedDeclaration instanceof ResolvedFieldDeclaration) {
    ResolvedFieldDeclaration resolvedField = (ResolvedFieldDeclaration) resolvedDeclaration;
    Optional<FieldDeclaration> fieldDeclaration = resolvedField.toAst();
}
```
Now we can do it by simple doing:
```java
ResolvedDeclaration resolvedDeclaration = ...;
Optional<FieldDeclaration> fieldDeclaration = resolvedDeclaration.toNode(FieldDeclaration.class);
```

This also allows add room for more complex examples that were not possible with the previous format, like looking for specific declarations. We can do it now by simply:
```java
Optional<VariableDeclarator> variableDeclarator = resolvedDeclaration.toNode(VariableDeclarator.class);
```

### CODE REVIEW

**Comment by 4everTheOne** [comment]
After the conversation with jlerbsc at https://github.com/javaparser/javaparser/issues/1940#issuecomment-773414969 that demonstrated the impact of this changes, decided to update this PR to reduce this impact. To reduce this impact a new AssociableToAst was created with a new name AssociatedWithAST and behavior exactly as described in the first comment.

Also AssociableToAst has been marked as Deprecated for not complying with the expected documentation.

**Comment by jlerbsc** [comment]
It would be possible to have a clearer PR by deleting all unnecessary commits.

File: javaparser-core/src/main/java/com/github/javaparser/resolution/declarations/AssociableToAST.java [hunk 1]
```diff
@@ -52,7 +51,23 @@
      * In these cases getWrappedNode is particularly nice because it returns the right type of AST node,
      * not just a Node.
      */
-    default Optional<N> toAst() {
-        throw new UnsupportedOperationException();
+    Optional<Node> toAst();
```
Comment (jlerbsc) [inline javaparser-core/src/main/java/com/github/javaparser/resolution/declarations/AssociableToAST.java]: Maybe a default implementation that returns Optional.empty() would simplify the code

**Comment by jlerbsc** [comment]
> After the conversation with jlerbsc at [#1940 (comment)](https://github.com/javaparser/javaparser/issues/1940#issuecomment-773414969) that demonstrated the impact of this changes, decided to update this PR to reduce this impact. To reduce this impact a new AssociableToAst was created with a new name AssociatedWithAST and behavior exactly as described in the first comment.
> 
> Also AssociableToAst has been marked as Deprecated for not complying with the expected documentation.

Given that the next version of JP will introduce a break in the API (this will be a major version certainly 4.0.0.), I send back my opinion, so that you can integrate this change directly into the existing interface.

**Comment by jlerbsc** [comment]
Did you complete your PR?

**Comment by 4everTheOne** [comment]
Yes I did. Sorry I forgot to leave a comment.

**Comment by jlerbsc** [comment]
Thank you for this contribution.

**Pull request merged**